### PR TITLE
3.next - Add RouteBuilder::loadPlugin.

### PR DESCRIPTION
**Pull request by markstory**
Deprecate the static method and provide a route builder method that allows routes to be loaded inside scopes.

Refs #10689

### CODE REVIEW

File: src/Routing/RouteBuilder.php [hunk 1]
```diff
@@ -519,6 +521,35 @@ protected function _methodRoute($method, $template, $target, $name)
     }
 
     /**
+     * Load routes from a plugin.
+     *
+     * The routes file will have a local variable named `$routes` made available which contains
```
Comment (ravage84) [inline src/Routing/RouteBuilder.php]: will -> must?

File: src/Routing/RouteBuilder.php [hunk 2]
```diff
@@ -519,6 +521,35 @@ protected function _methodRoute($method, $template, $target, $name)
...
+     * the current RouteBuilder instance.
+     *
+     * @param string $name The plugin name
+     * @param string $file The routes file to load. Defaults to `routes.php`
```
Comment (ravage84) [inline src/Routing/RouteBuilder.php]: 👍

File: src/Routing/RouteBuilder.php [hunk 3]
```diff
@@ -519,6 +521,35 @@ protected function _methodRoute($method, $template, $target, $name)
...
+     * @return void
+     * @throws \Cake\Core\Exception\MissingPluginException When the plugin has not been loaded.
+     * @throws \InvalidArgumentException When the plugin does not have a routes file.
+     */
+    public function loadPlugin($name, $file = 'routes.php')
+    {
+        if (!Plugin::loaded($name)) {
+            throw new MissingPluginException(['plugin' => $name]);
+        }
+        $path = Plugin::configPath($name) . DIRECTORY_SEPARATOR . $file;
```
Comment (ravage84) [inline src/Routing/RouteBuilder.php]: I'd insert a blank line before for readability.

File: src/Routing/RouteBuilder.php [hunk 4]
```diff
@@ -519,6 +521,35 @@ protected function _methodRoute($method, $template, $target, $name)
...
+        if (!file_exists($path)) {
+            throw new InvalidArgumentException(sprintf(
+                'Cannot load routes for the plugin named %s. The %s file does not exist.',
+                $name,
+                $path
+            ));
+        }
+        $routes = $this;
```
Comment (ravage84) [inline src/Routing/RouteBuilder.php]: I'd insert a blank line before for readability.

**Comment by ravage84** [comment]
@markstory can you elaborate what the implications are for the future app setup?

File: src/Routing/RouteBuilder.php (hunk 1 above)
Comment (markstory) [inline src/Routing/RouteBuilder.php]: I was trying to say the routes file will have access to `$routes` as a local variable.

**Comment by markstory** [comment]
@ravage84 I would like to see `Plugin::routes()` go away, and instead plugins either use events to automagically add their routes, or the installation of plugins requires adding `$routes->loadPlugin(...)` in the right place.

File: src/Routing/RouteBuilder.php (hunk 1 above)
Comment (ravage84) [inline src/Routing/RouteBuilder.php]: Yeah, I see now. Never mind.

**Comment by ravage84** [comment]
@markstory I see. I just "fear" the installation of plugins or better said the configuration of plugin routes could become more complicated because of this...

**Comment by markstory** [comment]
@ravage84 Fully removing `Plugin::routes()` and only having this method would make plugin installation more complex, which is not my goal. I want to make routing easier to understand and simpler to change. I found that `Plugin::routes()` is quite opaque and makes routing harder to understand and change.

Perhaps `Plugin::routes()` can live alongside this new method as a simple way to splat all the routes at once, while this new method gives people finer grained control over routes.

**Comment by lorenzo** [comment]
@markstory what happens if you use both `loadPlugin` and `Plugin::routes()`. Should this be an error condition?

**Comment by markstory** [comment]
@lorenzo The routes would be loaded twice. Which could result in incorrect URLs being generated. While I like to use `include_once` for routes to prevent that, but that complicates testing as we need to reload routes quite often.

If the routes are named, they will hit an error, but un-named routes are harder to catch.

**Pull request merged**